demo/mindstorms_widgets.py

```diff
@@ -0,0 +1,321 @@
+
+import sys, time
+from ev3dev import *
+# The mindstorms_widgets class is a wrapper for the python-ev3dev
+# language binding.  It implements interfaces that correspond directly
+# to the action and sensor programming blocks (widgets) from the 
+# mindstorms LabVIEW software.  This allows leverage from the amazing
+# documentation already provided by Lego.  Some exceptions are:
+# * It doesn't auto-detect the devices.  You need to call
+#   connect_motor or connect_sensor first for each device you use
+# * There is an assumption on motor connections
+#   'medium' motor must be connected to Port A
+#   'right' large motor must be connected to Port B
+#   'left'  large motor must be connected to Port C
+#   These names are used in the interface instead of the port names
+# * degrees and rotations inputs must always be positive
+#   use negative power to go backward
+# * Rather than having a separate unregulated_motor, regulated is an
+#   an optional input parameter (default = True)
+# * The direction input to move_steering doesn't behave the same 
+#   because it supports rotation i.e. negative direction for one motor
+# * motor_rotation_measure, 'current_power' returns both the duty_cycle_sp
+#   (the actual power level being applied to the motor) and speed_sp
+# * sound includes an option to give a sentence and have the robot speak
+
+# Known issues/limitations
+# * no error handling/recovery
+# * volume for tones doesn't work very well (seems to have only 2 levels)
+# * pulse option for brick_status_light doesn't work
+# * when reading the heading and proximity in beacon mode, ir_sensor_measure
+#   returns good values at first but continuing to poll it seems to result
+#   in inconsistent values
+# * ultrasonic and gyro sensors are not implemented
+
+# Mindstorms uses the following coefficients to compute speed_sp
+# based on a power input in range 0-100 i.e. speed_sp = x * power
+# medium motor: x = 16.4
+# large motor : x = 10.3
+# e.g. power of 10 sets medium motor speed_sp to 160 (deg/sec)
+
+MEDIUM_MOTOR_SPEED_COEFFICIENT = 16.4
+LARGE_MOTOR_SPEED_COEFFICIENT  = 10.3
+
+class mindstorms_widgets:
+    def __init__(self):
+        self.motor = {
+            'medium':None,
+            'left'  :None,
+            'right' :None
+        }
+
+    def _set_motor_attribs( self, size, motors, powers,
+                            invert, regulated, brake_at_end ):
+        # For regulated speed, speed_sp is tacho counts per 
+        # second which for lego EV3 motors is also degrees per second.
+        # To provide a mindstorms 0-100 power value for regulated
+        # speed we need to convert it.
+        speed_x = LARGE_MOTOR_SPEED_COEFFICIENT
+        if size == 'medium':
+            speed_x = MEDIUM_MOTOR_SPEED_COEFFICIENT
+            
+        for m,p in zip ( motors, powers ):
+            # Negative power and invert are synonomous, so the easiest
+            # logic is to invert the polarity when negative power is requested

```

Comment:
So you are saying it is surprising to the user if they put a negative power but I convert it to a positive power and toggle the invert attribute for them... I can see that.  This trick avoided some other complexity related to using the steering function but I can look into fixing that a different way.